ec2_vpc_subnet: Add IPv6 support

shepdelacreme opened the pull request:
##### SUMMARY
Adding support for creating/modifying VPC subnets with IPv6 CIDR blocks

Details about adding IPv6 support to AWS VPC modules can be found in #27800 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_vpc_subnet


##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.5.0 (ipv6_support_for_ec2_vpc_subnet 6c743e658b) last updated 2017/09/15 17:19:29 (GMT -400)
  config file = /Users/user/.ansible.cfg
  configured module search path = [u'/Users/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/user/devel/ansible/lib/ansible
  executable location = /Users/user/devel/ansible/bin/ansible
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
wrap boto3 connection in try/except

update module documentation and add RETURN docs

add IPv6 support to VPC subnet module

rename ipv6cidr to ipv6_cidr, use required_if for parameter testing, update some failure messages to be more descriptive

DryRun mode was removed from this function a while ago but exception handling was still checking for it, removed

add wait and timeout for subnet creation process

fixup the ipv6 cidr disassociation logic a bit per review

update RETURN values per review

added module parameter check

removed DryRun parameter from boto3 call since it would always be false here

fix subnet wait loop

add a purge_tags parameter, fix the ensure_tags function, update to use compare_aws_tags func

fix tags type error per review

remove **kwargs use in create_subnet function per review

willthames commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -228,38 +421,49 @@ def main():
-            az=dict(default=None, required=False),
+            az=dict(default='', required=False),
             cidr=dict(default=None, required=True),
+            ipv6cidr=dict(default='', required=False),
```
Prefer ipv6_cidr

willthames replied:
This will by default delete IPv6 CIDRs, according to your example documentation

shepdelacreme replied:
What is the desired behavior for modules like this since users can assign and unassign IPv6 CIDR blocks to subnets at will. I figured this behavior made the most sense. This way if you have your subnets defined as a list in your vars file and you want to change or remove the CIDR you just make that change in the subnets list.

The other way would be to have a `purge_ipv6_cidr` parameter or to differentiate between `ipv6cidr is None and ipv6cidr is ''`?

willthames commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -228,38 +421,49 @@ def main():
 
     module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
 
+    if module.params.get('assign_ipv6_on_create') and not module.params.get('ipv6cidr'):
```
This should probably just be `required_if`

shepdelacreme replied:
I initially attempted to get this working using required_if but could not and I could not find an example of using required_if to test against a boolean parameter. i.e. What I need is: `if assign_ipv6_on_create is True then ipv6cidr is a required value`

I tried this per something similar I found in one of the Azure modules but it did not work.
```python
    required_if = [('assign_ipv6_on_create', True, ['ipv6cidr'])]

    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True, required_if=required_if)
```

shepdelacreme replied:
Nevermind I figured this one out.

```python
required_if = [('assign_ipv6_on_create', True, ['ipv6cidr'])]
```

willthames commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -172,38 +296,107 @@ def ensure_map_public(conn, module, subnet, map_public, check_mode):
+    try:
+        conn.modify_subnet_attribute(SubnetId=subnet['id'], AssignIpv6AddressOnCreation={'Value': assign_ipv6_on_create})
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg=e.message, exception=traceback.format_exc(),
```
State what you're trying to do when failing
```
module.fail_json(msg="Couldn't set assign-ipv6-on-creation: " + e.message, ...)
```

willthames commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -172,38 +296,107 @@ def ensure_map_public(conn, module, subnet, map_public, check_mode):
+            conn.disassociate_subnet_cidr_block(AssociationId=subnet['ipv6_association_id'])
+        except botocore.exceptions.ClientError as e:
+            # set assign ipv6 on creation value back to original value before failing
+            ensure_assign_ipv6_on_create(conn, module, subnet, subnet['assign_ipv6_address_on_creation'], check_mode)
```
If both of these operations fail, it's going to be unclear what actually happened.

shepdelacreme replied:
I see what you are saying. The problem I have is that I have to turn off the `assign_ipv6_address_on_create` attribute before disassociating an IPv6 CIDR block and then set that attribute back to its intended value after assigning a new IPv6 CIDR. I was trying to avoid leaving the subnet in an unintended state on failure. i.e. `assign_ipv6` attribute gets disabled but then the disassociate operation fails for some reason.

I'm not sure if I, should be trying to correct from that failure, just provide more detailed error messages about the potentially bad state, move the `ensure_assign_ipv6_on_create` operation to the `try` block or to a `finally` block?

willthames commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -77,8 +90,107 @@
+'''
+
+RETURN = '''
+subnet.id:
```
This should `subnet` with type `complex` and `contains:` the rest of the information

shepdelacreme commented:
ready_for_review

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -77,8 +108,112 @@
+subnet:
+    description: Dictionary of subnet values
+    returned: I(state=present)
+    type: complex
```
Is the docsite bug that causes strangeness when rendering complex types resolved? I don't recall the issue number, but I think @willthames pointed it out.

shepdelacreme replied:
Hmm ok. I had it in a different format i.e `subnet.id` but @willthames requested it changed to this format. Is there a way for me to test the resulting docsite output? i.e Build it locally?

willthames replied:
```
cd docs/docsite; MODULES=ec2_vpc_subnet make docs
```

shepdelacreme replied:
This is what the docs look like with the complex types. Let me know if this is acceptable.

<img width="1113" alt="screen shot 2017-10-19 at 2 28 25 pm" src="https://user-images.githubusercontent.com/5123728/31787492-f5143300-b4d9-11e7-9176-f2e1d0436072.png">

ryansb replied:
Ah, great, that looks like we fixed complex representation!

willthames replied:
Although for some reason `I(state=present)` isn't being correctly rendered. Could just change that to when state is present.

I won't block on this, it seems more like a doc generation bug

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -102,64 +237,115 @@ def get_subnet_info(subnet):
+        subnet['tags'] = boto3_tag_list_to_ansible_dict(subnet['tags'], 'key', 'value')
     else:
-        subnet['tags'] = dict()
+        subnet['tags'] = [dict()]
```
I think the `tags` default was correct as `dict()`. Is there a reason it needs to be switched to a list-of-dict when there are no tags?

shepdelacreme replied:
Mistake on my part I believe. I was having issues with `boto3_tag_list_to_ansible_dict` returning the correct info which is resolved now.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -102,64 +237,115 @@ def get_subnet_info(subnet):
+    return False
+
+
+def create_subnet(conn, module, **kwargs):
```
Moving to `**kwargs` makes this function more opaque, since by looking you can't tell what's used and what's not. Can you put back the mandatory kwargs and make any optional ones defaulted?

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -45,15 +52,39 @@
     required: false
     default: false
     version_added: "2.4"
+  assign_ipv6_on_create:
```
I'm not big on this name, since it's not clear on creation of what. Maybe `assign_instances_ipv6` or similar would be clearer?

Commit:
Add integration test suite for ec2_vpc_subnet

willthames commented:
We now have a [policy](https://github.com/ansible/community/blob/master/group-aws/integration.md#policy) that feature fixes to stableinterface modules must be accompanied by tests. I actually created a smaller [PR for ec2_vpc_subnet](https://github.com/ansible/ansible/pull/31870) that comes with such tests. The question is whether you'd prefer to rebase onto my PR with tests (and update tests for the features you're adding).

As far as I can tell, @ryansb's request for a change of name of assign_ipv6_on_create is the only really outstanding item

shepdelacreme commented:
@willthames Great! I'll attempt to rebase on your test PR and fix up anything that doesn't work from there.

willthames commented:
@shepdelacreme thanks! I've just added the test and fix for #31905 too.

Commit:
rebased on #31870, fixed merge conflicts, and updated error messages

Commit:
fixes to pass tests

Commit:
add test for failure on invalid ipv6 block and update tags test for purge_tags=true function

Commit:
fix pylint issue

shepdelacreme commented:
Ok rebased on @willthames PR with tests for this module and got it passing.

I updated the remove tags test to use the `purge_tags` feature on the new module per the note that was left in the test playbook. This changes the behavior of the module slightly but is in line with how other modules operate. i.e. `purge_tags` defaults to `True`. 

I added another single test that tries to add an invalid ipv6 cidr block to the subnet and checks for failure. 

The main problem I've run into now is that I realize I can't actually put tests in the IPv6 block stuff  until the `ec2_vpc_net` module supports creating a VPC with an IPv6 cidr. If the aws cli was available I could use that or I could write a basic script that uses boto3 to do it in the meantime but not sure if that would be considered too brittle. Let me know what the preferred path here is.

shepdelacreme commented:
hmm looks like I hit a perm issue with the test IAM policy...but only in the `ubuntu1604py3` docker image. I'm trying to get a copy of the testing policy so I can validate locally what is happening.

Commit:
fix exception handling error when run with python3

Commit:
add ipv6 tests and fix module code

Commit:
Add permissions to hacking/aws_config/testing_policies/ec2-policy.json for adding IPv6 cidr blocks to VPC and subnets

shepdelacreme commented:
This is ready_for_review ...all tests should pass once the CI IAM permissions are updated.

willthames commented on test/integration/targets/ec2_vpc_subnet/tasks/main.yml:
```diff
@@ -0,0 +1,376 @@
+        state: present
+      register: vpc_subnet_recreate
+
+    - name: assert recreation changed nothing (expected changed=true)
```
changed=false

willthames commented on test/integration/targets/ec2_vpc_subnet/tasks/main.yml:
```diff
@@ -0,0 +1,376 @@
+      assert:
+        that:
+           - 'not vpc_subnet_recreate.changed'
+           - 'vpc_subnet_recreate.subnet.id.startswith("subnet-")'
```
more useful to do vpc_subnet_recreate.subnet_id == vpc_subnet_create.subnet_id

shepdelacreme replied:
What if I did `vpc_subnet_recreate.subnet == vpc_subnet_create.subnet` and compared the entire complex returned value? I'm not sure how Ansible would handle a comparison like this, however it works when I tried it.

Commit:
fix type in tests and update assert conditional to check entire returned value

mattclay commented on test/integration/targets/ec2_vpc_subnet/tasks/main.yml:
```diff
@@ -0,0 +1,370 @@
+      environment:
+          AWS_ACCESS_KEY_ID: '{{aws_access_key}}'
+          AWS_SECRET_ACCESS_KEY: '{{aws_secret_key}}'
+          #AWS_SESSION_TOKEN: '{{security_token}}'
```
CI failure due to lack of the `AWS_SESSION_TOKEN` environment variable. You'll need to make this task set the environment variable if it's available and omit it otherwise.

shepdelacreme replied:
I was getting local failures originally when I included `AWS_SESSION_TOKEN` but that is fixed now. I think its failing now because the CI IAM permissions.

Commit:
add AWS_SESSION_TOKEN into environment for aws cli commands to work in CI

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -102,167 +237,292 @@ def get_subnet_info(subnet):
+        timer = time.time() + wait_timeout
+        while time.time() < timer:
+            time.sleep(pause)
+            subnet = subnet_exists(conn, module, subnet['id'])
```
Can you use the waiter for this instead? http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Waiter.SubnetAvailable

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -102,167 +237,292 @@ def get_subnet_info(subnet):
 
     if 'tags' in subnet:
-        subnet['tags'] = boto3_tag_list_to_ansible_dict(subnet['tags'])
+        subnet['tags'] = boto3_tag_list_to_ansible_dict(subnet['tags'], 'key', 'value')
```
You can remove 'key' and 'value' since backwards compatibility was fixed for this in 56fe9499d14fa6e238fd3f32b1b33e6e7524e753.

Commit:
remove key and value options from call to boto3_tag_list_to_ansible_dict

Commit:
remove wait loop and use boto3 EC2 waiter

shepdelacreme commented:
Updated based on @s-hertel's review. CI failed but it doesn't look related to the module tests.

Failed Shippable jobs showing:
```
genExec|_common|saveState|_postFiles, :postFilesByResourceId failed for resourceId: 26365 with error Saving clone script failed for id: 26365
```

s-hertel commented:
Reran shippable and all passed.

@ryansb @willthames Can you review this again, please? I think your requested changes have been addressed.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -110,159 +245,266 @@ def get_subnet_info(subnet):
-            az=dict(default=None, required=False),
+            az=dict(default='', required=False),
             cidr=dict(default=None, required=True),
+            ipv6_cidr=dict(default='', required=False),
```
Is there ever a case where we'll want to distinguish between "user intentionally set to empty" vs. "user omitted the v6 option"?

shepdelacreme replied:
I wrote this with the opinion that you would need to be specific with setting this value. It made more sense to me and my use of the module. I wouldn't use it but I'll gladly add a "purge_ipv6" option or something if that is the consensus.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -110,159 +245,266 @@ def get_subnet_info(subnet):
     argument_spec.update(
         dict(
-            az=dict(default=None, required=False),
+            az=dict(default='', required=False),
```
I don't see why this should be changed, it seems like the v6 support shouldn't change this default.

shepdelacreme replied:
This fixes #31905

ryansb replied:
I see - if possible I'd like this in a separate PR because it's not related to IPv6, and has its own bug open already.

shepdelacreme replied:
I'll back this out to `default=None` and see if things still work. I believe the change that @willthames made in #31870 fixes this issue without changing the argument default and I've rebased on top of that PR so his changes exist in my branch as well.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -110,159 +245,266 @@ def get_subnet_info(subnet):
 
-    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+    if module.params.get('assign_instances_ipv6') and not module.params.get('ipv6_cidr'):
+        module.fail_json(msg="assign_instances_ipv6 is True but ipv6_cidr is None or an empty string")
```
Technically this will never be `None` because of the defaults.

shepdelacreme replied:
I can remove "None" from the message but you CAN cause the value to be None/null in the module by doing something stupid like this:
```yaml
    - name: create subnet
      ec2_vpc_subnet:
        cidr: "10.232.232.128/28"
        vpc_id: vpc-123456
        ipv6_cidr:
        assign_instances_ipv6: true
        state: present
```
Let me know if the preference is to change the error message.

ryansb replied:
I thought the default was `''` so a `None` would be overridden in the default-setting?

shepdelacreme replied:
Argument defaults only take effect if you don't specify a parameter at all. However if you do something boneheaded and specify a parameter with no value...i.e `ipv6_cidr: ` then it won't use the default setting for the arg and you get `None` as the value.

ryansb commented on test/integration/targets/ec2_vpc_subnet/tasks/main.yml:
```diff
@@ -0,0 +1,370 @@
+---
```
These look great! Thanks for adding

ryansb commented on test/integration/targets/ec2_vpc_subnet/tasks/main.yml:
```diff
@@ -0,0 +1,370 @@
+          AWS_SECRET_ACCESS_KEY: '{{aws_secret_key}}'
+          AWS_SESSION_TOKEN: '{{security_token}}'
+          AWS_DEFAULT_REGION: '{{ec2_region}}'
+      register: result
```
This is never used, so I don't think we should register this

shepdelacreme replied:
Thanks! Removed this and another unused `register: result` in https://github.com/ansible/ansible/pull/30444/commits/b4f4f3ec6a49c379eba237fc34e97a2efbd26020

Commit:
remove unused register: result vars

Commit:
revert az argument default value to original setting default=None

ryansb commented:
Rebased. Now to re-shippable.
